sdks/python/apache_beam/io/iobase.py

```diff
@@ -659,6 +675,167 @@ def fraction_consumed(self):
+      unfinished read itself represents a unfinished split point. This
+      condition holds independent of whether the input is splittable.
+    * A finished read, for which RangeTracker.set_done() has been invoked,
+      remaining number of split points is 0. This condition holds independent of
+      whether the input is splittable.
+    * For example 1: After returning record #30 (starting at 1) out of 50 in a
+      perfectly splittable 50-record input, remaining number of split points
+      should be 21 (20 remaining + 1 current) if the total number of records is
+      known.
+    * For example 2: After returning a record in block 3 in a block-compressed
+      file consisting of 5 blocks, remaining number of split points should be 3
+      (since blocks 4 and 5 can be processed in parallel by new source read
+      invocations produced via dynamic work rebalancing, while the current
+      source read continues processing block 3) if the total number of blocks is
+      known.
+    * For example (3): a source read for any non-empty unsplittable input,
+      remaining number of split points is 1 until ``RangeTracker.set_done()`` is
+      invoked, at which point remaining number of split points is 0.
+    * For any source read: After ``RangeTracker.try_claim()`` is invoked for the
+      last split point in a file (e.g., the last record in example (1), the
+      first record in the last block for example (2), or the first record in
+      the file for example (3), number of remaining split points is 1, which
+      represents the current source read.
+
+    By default ``RangeTracker` returns ``RangeTracker.SPLIT_POINTS_UNKNOWN`` for
+    both consumed and remaining number of split points, which indicates that the
+    number of split points consumed and remaining is unknown.
+
+    Returns:
+      a pair that gives consumed and remaining number of split points. Each
+      value should be an integer that is larger than or equal to zero or
+      ``RangeTracker.SPLIT_POINTS_UNKNOWN``.
+    """
+    return (RangeTracker.SPLIT_POINTS_UNKNOWN,
+            RangeTracker.SPLIT_POINTS_UNKNOWN)
+
+  def set_split_points_remaining_callback(self, callback):
+    """Sets a callback function for determining split points remaining.
+
+    By invoking this function, a ``BoundedSource`` can set a callback function
+    that may get invoked by the RangeTracker to determine the number of
+    remaining (unconsumed) split points. The callback function accepts a single
+    parameter, a stop position for the BoundedSource (stop_position). If the
+    record currently being consumed by the ``BoundedSource`` is at position
+    current_position, callback should return the number of split points within
+    the range [current_position, stop_position). Note that this should include
+    the split point currently being consumed by the source.
+
+    The given callback must only be invoked after acquiring the lock of the
```

Comment:
It seems, however, that care still needs to be taken as this may be called from a different thread than the one that set the callback...